### PR TITLE
[TSE] Implement Exchange Service

### LINKED ISSUE #4: [TSE] Implement the Exchange Service
### Description
This service will be responsible for interacting with all the supported exchanges, utilizing the [CCXT](https://docs.ccxt.com/#/) library.

**Important!** Every exchange operation that affects the exchange state(ex. create order, cancel order) should be saved in the database (Exchange Operation Service).

### To Do
- Implement `executeMarketTrade` method
- Implement `executeLimitTrade` method
- Implement `cancelOrder` method

### Deliverables
- [ ] All tasks completed
- [ ] Unit tests
- [ ] Documentation

### PR DESCRIPTION
**Pull request by MWBlocky**
## Description

This Pull Request introduces the implementation of trade execution and order management methods within the Exchange Trade Service, using the CCXT library for interaction with supported exchanges. The service ensures that every exchange operation affecting the exchange state, such as order creation or cancellation, is saved in the database via the Exchange Operation Service.

## Summary of changes

- **Implemented `executeMarketTrade` method**: Executes a market trade for a specified symbol and amount on a given exchange.
- **Implemented `executeLimitTrade` method**: Executes a limit trade with a specified price and amount on a given exchange.
- **Implemented `cancelOrder` method**: Cancels an existing order based on the provided order ID and symbol on a given exchange.
- **Integrated database logging**: Ensured that each trade operation (market trade, limit trade, and order cancellation) is logged to the database through the Exchange Operation Service.
- **Added unit tests**: Verified the correct functionality of all methods, ensuring they handle various edge cases and errors appropriately.
- **Documentation**: Updated the service documentation to include details on the new methods, their usage, and examples.

## Related issues

Fixes #4

### CODE REVIEW

File: packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts [hunk 1]
```diff
@@ -21,7 +21,7 @@ export class CreateLimitOrderCommand extends CreateMarketOrderCommand {
 export type OrderCommand = CreateMarketOrderCommand | CreateLimitOrderCommand;
 
 export class ExchangeOperationCommand {
-  orderEntityId: number;
+  orderEntityId?: number | undefined;
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts:24]: Why can:  `orderEntityId?` be undefined?
In which scenario?

File: packages/tse-backend/src/modules/exchange-operation/order.service.ts [hunk 2]
```diff
@@ -17,7 +17,9 @@ export class OrderService {
 
   async persistOrderActivity(command: ExchangeOperationCommand): Promise<void> {
     const { orderEntityId, status, orderId, details } = command;
-    const order = await this.repository.findById(orderEntityId);
+    const order = orderEntityId
+      ? await this.repository.findById(orderEntityId)
+      : await this.repository.findByOrderId(orderId);
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-operation/order.service.ts:22]: What's the difference between `orderEntityId` and `orderId`?

File: packages/tse-backend/src/modules/exchange-trade/model/exchange-trade.model.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  @IsString()
+  orderId: string;
+
+  @AutoMap()
+  @ApiProperty()
+  @IsString()
+  symbol: string;
+}
+
+export class CancelOrderCommand {
```
Comment (piotrswierzy) [inline packages/tse-backend/src/modules/exchange-trade/model/exchange-trade.model.ts:92]: Use `interface` instead of `class` if you want to create a wrapper:
https://stackoverflow.com/questions/51716808/when-use-a-interface-or-class-in-typescript

and try not to extend classes, "Favour Composition over Inheritance:
https://en.wikipedia.org/wiki/Composition_over_inheritance"

Comment (piotrswierzy) [reply packages/tse-backend/src/modules/exchange-trade/model/exchange-trade.model.ts:92]: it's for all wrappers

**Review by piotrswierzy — CHANGES_REQUESTED** [review]

File: packages/tse-backend/src/modules/exchange-operation/order.service.ts (hunk 2 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-operation/order.service.ts:22]: `orderEntityId` is `id` from `packages/tse-backend/src/common/entities/order.entity.ts`
`orderId` is `result.id` from `packages/tse-backend/src/modules/exchange-trade/exchange-trade.service.ts`:
```
const result = await exchangeInstance.createOrder(
        command.symbol,
        MarketOrderType.MARKET_ORDER,
        command.side,
        command.amount,
      );
      this.logger.log(
        `Market trade executed successfully: ${JSON.stringify(result)}`,
      );
      await this.saveExchangeOperation(
        savedData.id,
        OrderStatus.EXECUTED,
        result.id,
        result,
      );
```

I will rename orderId into orderExternalId to clarify.

File: packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts (hunk 1 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts:24]: `packages/tse-backend/src/modules/exchange-trade/exchange-trade.service.ts`
The `cancelOrder` method uses `orderId` to cancel the order, and to update the order status we don't need orderEntityId, because we get the entity using `orderId`.

Comment (piotrswierzy) [reply packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts:24]: Then maybe it's better to create a separate DTO for canceling?

File: packages/tse-backend/src/modules/exchange-trade/model/exchange-trade.model.ts (hunk 3 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-trade/model/exchange-trade.model.ts:92]: In NestJS for DTO we need to use classes to get the benefits of validation, transformation and documentation features.
For Commands decorator @AutoMap() doesn’t work with interfaces..

File: packages/tse-backend/src/modules/exchange-operation/order.service.ts (hunk 2 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-operation/order.service.ts]: fix: b51ceb05

File: packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts (hunk 1 above)
Comment (MWBlocky) [reply packages/tse-backend/src/modules/exchange-operation/model/exchange-operation.model.ts]: Fixed: 0b747d9f
